build_tools/cmake/iree_check_test.cmake

```diff
@@ -238,6 +238,11 @@ function(iree_check_single_backend_test_suite)
     # such as when cross compiling, then we can't easily check for which
     # compiler target backends are enabled. Just assume all are enabled and only
     # rely on the runtime HAL driver check above for filtering.
+
+    # HACK: Always skip WebGPU when cross compiling until it has a HAL driver.
```

Comment:
could we make this not run the tests if DRIVER is omitted? I see a note about it above but am unclear why we'd allow not passing DRIVER to still run.

Reply:
Soooo....

We already skip running tests if there is no DRIVER, and that's what WebAssembly is relying on today: https://github.com/iree-org/iree/blob/51d4dc6828ee60376e7ee630d3a5e9ef96a2db8e/tests/e2e/xla_ops/BUILD#L393-L402

This gets tricky with cross-compilation though. We don't currently have a way to know from the cross-compilation CMake environment which compiler targets were enabled in the host tools build. Since the WebGPU _compiler target_ is not enabled by default, one of our Android builds was trying to build the WebGPU tests. Thinking about this a bit more, I can tweak this hack to skip all tests that omit `DRIVER` when cross compiling. The real fix is mentioned in the PR description though - stop trying to use `iree_check_test` for just testing compilation. It's much easier to edit `iree_check_test` than it is to add a new test type though :/

![image](https://user-images.githubusercontent.com/4010439/204402315-d6a50259-4910-47f2-b1a4-601030faa21a.png)

Reply:
@GMNGeoffrey will you forgive me for the hack in iree_check_test? 😅 
(I'll wait a bit before merging to give some time if you'd like to review that part)

Reply:
Eeeew. I'm pretty sure we're just not doing cross compiling correctly... We should be able to query the iree compiler tool to see what backends it supports, shouldn't we? I guess that opens us up to silent failures when the tool isn't compiled as expected though. In CMake you can actually have tests depend on other tests. I've been thinking for a while that that's actually how we should do the compilation tests rather than having them as part of the build system. That would resolve the weird `iree-test-deps` target too. But that is for another time. I guess this isn't the worst.

Reply:
Can you at least add a TODO and file an issue (or attach it to an existing issue)? "Rewrite absolutely all of our tests". "good first issue"

Reply:
> In CMake you can actually have tests depend on other tests. I've been thinking for a while that that's actually how we should do the compilation tests rather than having them as part of the build system.

Yeah, that would be great. I can file an issue for that and add it to my backlog (or someone else can tackle it :D)

Reply:
Filed https://github.com/iree-org/iree/issues/11354